src/functional/ffront/foast_passes/type_deduction.py

```diff
@@ -0,0 +1,95 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2014-2021, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+import warnings
+from typing import Optional
+
+import functional.ffront.field_operator_ast as foast
+from eve import NodeTranslator
+
+
+def is_complete_type(fo_type: foast.DataType) -> bool:
+    if fo_type is None:
+        return False
+    elif isinstance(fo_type, foast.DeferredSymbolType):
+        return False
+    elif isinstance(fo_type, foast.SymbolType):
+        return True
+    return False
+
+
+class FieldOperatorTypeDeduction(NodeTranslator):
+    """Deduce and check types of FOAST expressions and symbols."""
+
+    @classmethod
+    def apply(cls, node: foast.FieldOperator) -> foast.FieldOperator:
+        return cls().visit(node)
+
+    def visit_FieldOperator(self, node: foast.FieldOperator, **kwargs) -> foast.FieldOperator:
+        kwargs["symtable"] = kwargs.get("symtable", node.symtable_)
+        return foast.FieldOperator(
+            id=node.id,
+            params=self.visit(node.params, **kwargs),
+            body=self.visit(node.body, **kwargs),
+            location=node.location,
+        )
+
+    def visit_Name(self, node: foast.Name, **kwargs) -> foast.Name:
+        symtable = kwargs.get("symtable", {})
+        if node.id not in symtable:
+            warnings.warn(  # TODO (ricoh): raise this instead (requires externals)
+                FieldOperatorTypeDeductionError.from_foast_node(
+                    node, msg=f"Undeclared symbol {node.id}"
+                )
+            )
+            return node
+
+        symbol = symtable[node.id]
+        return foast.Name(id=node.id, type=symbol.type, location=node.location)
```

Comment:
I'm not 100% sure but it looks like this Visitor should use the `SymbolTableTrait.symtable_merger` as `context`:

```suggestion
class FieldOperatorTypeDeduction(NodeTranslator):
    """Deduce and check types of FOAST expressions and symbols."""
    
    contexts = (SymbolTableTrait.symtable_merger,)
    
    @classmethod
    def apply(cls, node: foast.FieldOperator) -> foast.FieldOperator:
        return cls().visit(node)

    def visit_FieldOperator(self, node: foast.FieldOperator, **kwargs) -> foast.FieldOperator:
        return foast.FieldOperator(
            id=node.id,
            params=self.visit(node.params, **kwargs),
            body=self.visit(node.body, **kwargs),
            location=node.location,
        )

    def visit_Name(self, node: foast.Name, **kwargs) -> foast.Name:
        symtable = kwargs["symtable"]
        if node.id not in symtable:
            warnings.warn(  # TODO (ricoh): raise this instead (requires externals)
                FieldOperatorTypeDeductionError.from_foast_node(
                    node, msg=f"Undeclared symbol {node.id}"
                )
            )
            return node

        symbol = symtable[node.id]
        return foast.Name(id=node.id, type=symbol.type, location=node.location)
```